### PR TITLE
Add a click_and_move widget test helper

### PR DESCRIPTION
**Pull request by dstansby**
Addresses part of https://github.com/matplotlib/matplotlib/issues/21774.

### CODE REVIEW

**Review by ericpre — APPROVED** [review]
Looks good, could it also be used in the following cases?
- replace `_resize_rectangle` by `click_and_move`
- `test_rectangle_rotate`
- `test_rectange_add_remove_set`

File: lib/matplotlib/testing/widgets.py [hunk 1]
```diff
@@ -83,3 +83,31 @@ def do_event(tool, etype, button=1, xdata=0, ydata=0, key=None, step=1):
     event = mock_event(tool.ax, button, xdata, ydata, key, step)
     func = getattr(tool, etype)
     func(event)
+
+
+def click_and_move(tool, start, end, key=None):
```
Comment (timhoffm) [inline lib/matplotlib/testing/widgets.py:88]:
```suggestion
def click_and_drag(tool, start, end, key=None):
```
I think moving the mouse while holding a button is called "drag".

File: lib/matplotlib/testing/widgets.py [hunk 2]
```diff
@@ -83,3 +83,31 @@ def do_event(tool, etype, button=1, xdata=0, ydata=0, key=None, step=1):
...
+    """
+    Click at *start*, move to end*, and then relese the mouse.
```
Comment (timhoffm) [inline lib/matplotlib/testing/widgets.py:90]:
```suggestion
    Helper to simulate a mouse drag operation.
    
    Click at *start*, move to *end*, and then relese the mouse.
```

File: lib/matplotlib/testing/widgets.py [hunk 3]
```diff
@@ -83,3 +83,31 @@ def do_event(tool, etype, button=1, xdata=0, ydata=0, key=None, step=1):
...
+
+    Parameters
+    ----------
+    tool : matplotlib.widgets.RectangleSelector
```
Comment (timhoffm) [inline lib/matplotlib/testing/widgets.py:94]:
```suggestion
    tool : `matplotlib.widgets.RectangleSelector`
```
But is this really specific to `RectangleSelector`? If so maybe the function name should be `click_and_drag_rectangle` or `click_and_drag_selector`.

File: lib/matplotlib/testing/widgets.py [hunk 4]
```diff
@@ -83,3 +83,31 @@ def do_event(tool, etype, button=1, xdata=0, ydata=0, key=None, step=1):
...
+    start : [Number, Number]
```
Comment (timhoffm) [inline lib/matplotlib/testing/widgets.py:95]:
```suggestion
    start : (float, float)
```
would be our standard convention (https://matplotlib.org/stable/devel/documenting_mpl.html#parameter-type-descriptions).

File: lib/matplotlib/testing/widgets.py [hunk 5]
```diff
@@ -83,3 +83,31 @@ def do_event(tool, etype, button=1, xdata=0, ydata=0, key=None, step=1):
...
+        Starting point.
```
Comment (timhoffm) [inline lib/matplotlib/testing/widgets.py:96]: In which coordinate system? Same for *end*.

File: lib/matplotlib/testing/widgets.py [hunk 6]
```diff
@@ -83,3 +83,31 @@ def do_event(tool, etype, button=1, xdata=0, ydata=0, key=None, step=1):
...
+    end : [Number, Number]
+        End point.
+    key : None or str
+        The key pressed when the mouse event triggered (see also `.KeyEvent`).
```
Comment (timhoffm) [inline lib/matplotlib/testing/widgets.py:100]:
```suggestion
        An optional key that is pressed during the whole operation
        (see also `.KeyEvent`).
```

File: lib/matplotlib/tests/test_widgets.py [hunk 7]
```diff
@@ -81,9 +82,7 @@ def onselect(epress, erelease):
     tool = widgets.RectangleSelector(ax, onselect, interactive=True,
                                      drag_from_anywhere=drag_from_anywhere)
     # Create rectangle
-    do_event(tool, 'press', xdata=0, ydata=10, button=1)
-    do_event(tool, 'onmove', xdata=100, ydata=120, button=1)
-    do_event(tool, 'release', xdata=100, ydata=120, button=1)
+    click_and_move(tool, start=[0, 10], end=[100, 120])
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_widgets.py:85]: I'd prefer tuples over lists for specifying points.

Comment (dstansby) [reply lib/matplotlib/tests/test_widgets.py]: How blocking is this comment? It's a bit of a pain to change now, but I do agree with you.

File: lib/matplotlib/testing/widgets.py (hunk 3 above)
Comment (dstansby) [reply lib/matplotlib/testing/widgets.py]: Thanks for catching, it's not specific to RectangleSelector

File: lib/matplotlib/testing/widgets.py (hunk 4 above)
Comment (dstansby) [reply lib/matplotlib/testing/widgets.py]: Thanks

**Comment by dstansby** [comment]
@timhoffm Thanks 😃

File: lib/matplotlib/tests/test_widgets.py (hunk 7 above)
Comment (timhoffm) [reply lib/matplotlib/tests/test_widgets.py]: I took the liberty to push a commit changing all the lists to tuples. It's actually not very painful at all. Regex-Replace: 1min + 1min to check the individual replacements. :)

**Pull request merged**